activemodel/lib/active_model/type/helpers/time_value.rb

```diff
@@ -69,20 +69,36 @@ def new_time(year, mon, mday, hour, min, sec, microsec, offset = nil)
             \z
           /x
 
-          def fast_string_to_time(string)
-            return unless ISO_DATETIME =~ string
-
-            usec = $7.to_i
-            usec_len = $7&.length
-            if usec_len&.< 6
-              usec *= 10**(6 - usec_len)
+          if RUBY_VERSION >= "3.2"
+            def fast_string_to_time(string)
+              return unless ISO_DATETIME.match?(string)
+
+              if is_utc?
+                # XXX: Wrapping the Time object with Time.at because Time.new with `in:` in Ruby 3.2.0 used to return an invalid Time object
+                # see: https://bugs.ruby-lang.org/issues/19292
+                ::Time.at(::Time.new(string, in: "UTC"))
+              else
+                ::Time.new(string)
+              end
+            rescue ArgumentError
+              nil
             end
+          else
+            def fast_string_to_time(string)
+              return unless ISO_DATETIME =~ string
```

Comment:
```suggestion
              return unless ISO_DATETIME.match?(string)
```

While we're at it.

Reply:
No, in this case (RUBY < 3.2) we use the matched `$1` - `$9` soon below, and thus this has to be `=~`.

Reply:
🤦 indeed.

Reply:
I was about to voice my surprise that a regexp might be faster than `Time.iso8601(String)`, but turns out the stdlib method is using a similar regexp.